internal/app/grpcserver/impl.go

```diff
@@ -67,25 +71,193 @@ func (c *conformanceServiceServer) Unary(
 }
 
 func (c *conformanceServiceServer) ClientStream(
-	_ v1alpha1.ConformanceService_ClientStreamServer,
+	stream v1alpha1.ConformanceService_ClientStreamServer,
 ) error {
-	// TODO - Implement ClientStream
-	return status.Error(codes.Unimplemented, "method ClientStream not implemented")
+	var responseDefinition *v1alpha1.UnaryResponseDefinition
+	firstRecv := true
+	var reqs []*anypb.Any
+
+	for {
+		if err := stream.Context().Err(); err != nil {
+			return err
+		}
+		msg, err := stream.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return err
+		}
+		// If this is the first message received on the stream, save off the response definition we need to send
+		if firstRecv {
+			responseDefinition = msg.ResponseDefinition
+			firstRecv = false
+		}
+		// Record all the requests received
+		msgAsAny, err := asAny(msg)
+		if err != nil {
+			return err
+		}
+		reqs = append(reqs, msgAsAny)
+	}
+
+	if err := grpcutil.AddHeaderMetadata(stream.Context(), responseDefinition.ResponseHeaders); err != nil {
+		return err
+	}
+	if err := grpcutil.AddTrailerMetadata(stream.Context(), responseDefinition.ResponseTrailers); err != nil {
+		return err
+	}
```

Comment:
It is only necessary to set headers and trailers through the context for unary RPCs. For all stream RPCs, it is more clear to instead use the `SetHeader` and `SetTrailer` methods on `stream`.

Reply:
Gotcha. Changed all the streaming endpoints do use the `stream` methods. Also, since unary was now the only one using `AddHeader/TrailerMetadata`, I removed it from the `grpcutil` package and just inlined right in the unary endpoint.